### PR TITLE
Replace `reschedule` with new DSL endpoints and refactor task scheduler

### PR DESCRIPTION
**Pull request by Bidski**
There is bug in `Sync` which can allow two messages in the same `SyncGroup` to be emitted in one order but processed in the opposite order.

This PR removes the reschedule DSL endpoint and replaces it with two new DSL enpoints
1. `Group`
2. `Pool`

`Group` extends the original functionality of `Sync` to allow for `SyncGroup`s with a concurrency larger than 1.

`Pool` adds new functionality to allow the user to request that certain reactions run in thread pools that are not the main thread or the default thread pool. The `Always` DSL makes use of this new DSL endpoint by creating a single-threaded thread pool for each distinct `Always` reaction.

All tasks, regardless of which thread pool they are executed on, are now scheduled for execution in the one scheduler. This means that the priority queue that `Sync` uses has now been removed

---

There is still a bug which causes the test for `BaseClock` to SIGABRT on Mac. Valgrind's DRD tool on linux flags something to do with Direct emits and ReactionStatistics but the precise cause is still unknown.

However, this bug existed prior to this PR

### CODE REVIEW

File: src/util/ThreadPoolDescriptor.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "ThreadPoolDescriptor.hpp"
+
+namespace NUClear {
+namespace util {
+
+    const uint64_t ThreadPoolIDSource::MAIN_THREAD_POOL_ID    = 0;
+    const uint64_t ThreadPoolIDSource::DEFAULT_THREAD_POOL_ID = 1;
```
Comment (TrentHouliston) [inline src/util/ThreadPoolDescriptor.cpp:24]: I think these can go in the header if they are constexpr?
Unless there is a linking problem somewhere where they've been taken by reference.
In the right context you can assign them to a variable and then pass them in to avoid that too I think.

File: src/util/ThreadPoolDescriptor.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#define NUCLEAR_UTIL_THREADPOOL_HPP
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+
+namespace NUClear {
+namespace util {
+
+    struct ThreadPoolIDSource {
```
Comment (TrentHouliston) [inline src/util/ThreadPoolDescriptor.hpp:28]: Why separated for ThreadPool but joined for group?

File: tests/dsl/MainThread.cpp [hunk 3]
```diff
@@ -26,19 +26,20 @@ class TestReactor : public NUClear::Reactor {
...
         // Run a task without MainThread to make sure it isn't on the main thread
-        on<Trigger<int>>().then([this] {
+        on<Trigger<int>>().then("Non-MainThread reaction", [this] {
             // We shouldn't be on the main thread
             REQUIRE(NUClear::util::main_thread_id != std::this_thread::get_id());
 
             emit(std::make_unique<double>(1.1));
         });
 
         // Run a task with MainTHread and ensure that it is on the main thread
```
Comment (TrentHouliston) [inline tests/dsl/MainThread.cpp:36]:
```suggestion
        // Run a task with MainThread and ensure that it is on the main thread
```

File: tests/dsl/SingleSync.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            [](const Message& m) { values.push_back("Received value " + std::to_string(m.val)); });
+
+        on<Trigger<ShutdownOnIdle>, Priority::IDLE>().then([this] { powerplant.shutdown(); });
+
+        on<Startup>().then([this] {
+            values.clear();
+            emit(std::make_unique<Message>(0));
+            emit(std::make_unique<Message>(1));
+            emit(std::make_unique<Message>(2));
+            emit(std::make_unique<Message>(3));
```
Comment (TrentHouliston) [inline tests/dsl/SingleSync.cpp:49]: Still quite likely to pass when it shouldn't

Try emitting in a for loop for 100 or 1000 or something

File: tests/log/Log.cpp [hunk 5]
```diff
@@ -165,7 +165,6 @@ TEST_CASE("Testing the Log<>() function", "[api][log]") {
     expected_count += levels.size() * (levels.size() + 1) / 2;  // Direct reaction logs
     expected_count += levels.size() * (levels.size() + 1) / 2;  // Indirect reaction logs
     expected_count += levels.size() * levels.size();            // Non reaction logs
-    expected_count += 2 + levels.size();                        // Post shutdown logs
```
Comment (TrentHouliston) [inline tests/log/Log.cpp:168]: What happened to post shutdown logs?

File: CMakeLists.txt [hunk 6]
```diff
@@ -48,6 +48,40 @@ if(CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
   set(MASTER_PROJECT ON)
 endif()
 
+option(USE_VALGRIND "Add extra annotations for better valgrind handling" OFF)
```
Comment (TrentHouliston) [inline CMakeLists.txt:51]: If we were to introduce these, they should be in another PR
Keep them for now while debugging if you want but remove them before the final merge

File: src/dsl/fusion/GroupFusion.hpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            }
+        };
+
+        // Case where there are 2 or more words remaining
+        template <typename Word1, typename Word2, typename... WordN>
+        struct GroupFuser<std::tuple<Word1, Word2, WordN...>> {
+
+            template <typename DSL>
+            static inline void group(threading::Reaction& /*reaction*/) {
+                throw std::runtime_error("Can not be a member of more than one group");
```
Comment (TrentHouliston) [inline src/dsl/fusion/GroupFusion.hpp:87]: Did we want to be able to be a part of multiple groups?

File: src/extension/ChronoController.hpp [hunk 8]
```diff
@@ -37,8 +37,10 @@ namespace extension {
                 // Lock the mutex while we're doing stuff
                 const std::lock_guard<std::mutex> lock(mutex);
 
-                // Add our new task to the heap
-                tasks.push_back(*task);
+                // Add our new task to the heap if we are still running
+                if (running) {
```
Comment (TrentHouliston) [inline src/extension/ChronoController.hpp:41]: Would powerplants running flag work for this? or is there an advantage to having our own

File: src/extension/ChronoController.hpp [hunk 9]
```diff
@@ -67,62 +69,66 @@ namespace extension {
...
                 wait.notify_all();
             });
 
             on<Always, Priority::REALTIME>().then("Chrono Controller", [this] {
                 // Acquire the mutex lock so we can wait on it
                 std::unique_lock<std::mutex> lock(mutex);
 
-                // If we have tasks to do
-                if (!tasks.empty()) {
+                if (running) {
```
Comment (TrentHouliston) [inline src/extension/ChronoController.hpp:80]: Is this needed? Won't the always itself just stop running if running were false?
To reduce the diff this could just be
```cpp
if(running) {
    return;
}
```

**Review by TrentHouliston — CHANGES_REQUESTED** [review]
Review of everything except Always.hpp, TaskScheduler.cpp and TaskScheduler.hpp

File: src/dsl/word/Group.hpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+#include "../../threading/ReactionTask.hpp"
+#include "../../util/GroupDescriptor.hpp"
+
+namespace NUClear {
+namespace dsl {
+    namespace word {
+
+        template <typename GroupType>
+        struct Group {
```
[the remark this answers is not in the document]
Comment (Bidski) [reply src/dsl/word/Group.hpp:33]: `Sync` is implicitly a group with only 1 task running at a time, `Group` allows you to specify the concurrency to be more than 1

File: src/util/ThreadPoolDescriptor.cpp (hunk 1 above)
Comment (Bidski) [reply src/util/ThreadPoolDescriptor.cpp:24]: I moved them here because there are undefined references with some compilers

File: tests/dsl/SingleSync.cpp (hunk 4 above)
Comment (Bidski) [reply tests/dsl/SingleSync.cpp:49]: Emit 100 messages? Or run the powerplant 100 times?

File: tests/log/Log.cpp (hunk 5 above)
Comment (Bidski) [reply tests/log/Log.cpp:168]: logs only work if `powerplant != nullptr`, but powerplant destruction sets the singleton to `nullptr`, not sure why these ever worked tbh

File: src/dsl/fusion/GroupFusion.hpp (hunk 7 above)
Comment (Bidski) [reply src/dsl/fusion/GroupFusion.hpp:87]: Im not a fan of the idea, I can't visualise a viable use case for it, but if you are insistent and there are no other outstanding issues we can add that in a new PR

File: src/extension/ChronoController.hpp (hunk 8 above)
Comment (Bidski) [reply src/extension/ChronoController.hpp:41]: I added our own because IOController has its own :shrug:

File: src/extension/ChronoController.hpp (hunk 9 above)
Comment (Bidski) [reply src/extension/ChronoController.hpp:80]: I think I saw what I thought was the chrono controller continuing to run after shutdown, so I added this to force the point, but now that I think about it it may have been to do with the prolific always spam we saw before

File: src/dsl/word/Group.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply src/dsl/word/Group.hpp:33]: Looks like it's hard coded to 1 to me

https://github.com/Fastcode/NUClear/pull/64/files#diff-1c298e57e77793d223e8e17d56223cf176ec8da8fa6be3613f7320067543b935R47

And if not, sync should just inherit from group like Single does from Buffer

File: tests/log/Log.cpp (hunk 5 above)
Comment (TrentHouliston) [reply tests/log/Log.cpp:168]: Powerplant destruction sets them to nullptr, but these were on shutdown, not destruction. Powerplant destruction should happen when there is nothing left, no reactors everything has unbound etc. There should be nothing left to log when it goes nullptr

Comment (Bidski) [reply tests/log/Log.cpp:168]: No, these were post shutdown, as the comment says

Comment (Bidski) [reply tests/log/Log.cpp:168]: hmm hang on, i think I missed something

File: src/util/ThreadPoolDescriptor.cpp (hunk 1 above)
Comment (TrentHouliston) [reply src/util/ThreadPoolDescriptor.cpp:24]: even with constexpr?

File: tests/dsl/SingleSync.cpp (hunk 4 above)
Comment (TrentHouliston) [reply tests/dsl/SingleSync.cpp:49]: Emit 100 messages

File: src/util/CallbackGenerator.hpp [hunk 11]
```diff
@@ -44,9 +47,15 @@ namespace util {
...
-            , transients(std::make_shared<typename TransientDataElements<DSL>::type>()){};
+        // Don't use this constructor if F is of type CallbackGenerator
+        template <typename F,
+                  typename std::enable_if<
+                      !std::is_same<typename std::remove_reference<typename std::remove_cv<F>::type>::type,
+                                    CallbackGenerator>::value,
+                      bool>::type = true>
+        CallbackGenerator(F&& callback)
+            : callback(std::forward<F>(callback))
+            , transients(std::make_shared<typename TransientDataElements<DSL>::type>()) {}
```
[the remark this answers is not in the document]
Comment (Bidski) [reply src/util/CallbackGenerator.hpp:58]: It could be solved by adding `explicit` to this constructor instead of the templating mess, but gcc5 doesn't like having `explicit`

File: src/util/ThreadPoolDescriptor.cpp (hunk 1 above)
Comment (Bidski) [reply src/util/ThreadPoolDescriptor.cpp]: These used to be `static constexpr ....` with an inline initialiser in the header, certain compilers hated it and complained about undefined references

File: tests/log/Log.cpp (hunk 5 above)
Comment (Bidski) [reply tests/log/Log.cpp]: We stopped processing all submitted tasks (even direct/immediate tasks) after `powerplant.shutdown()` was called, so anything emitted after shutdown was ignored. This should be fixed now

File: src/dsl/word/Pool.hpp [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+         *  This DSL will cause the creation of a new thread pool with a specific number of threads allocated to it.
+         *
+         *  All tasks for this reaction will be queued to run on threads from this thread pool.
+         *
+         *  Tasks in the queue are ordered based on their priority level, then their task id.
+         *
+         *  When this DSL is not specified the default thread pool will be used. For tasks that need to run on the main
+         *  thread use MainThread.
+         *
+         *  For best use, this word should be fused with at least one other binding DSL word.
```
Comment (TrentHouliston) [inline src/dsl/word/Pool.hpp:47]: I think we can just drop this word sentence from everywhere. Everything needs a binding dsl word

File: src/dsl/word/Pool.hpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+         *
+         * @par Implements
+         *  pool
+         *
+         * @tparam PoolType
+         *  A struct that contains the details of the thread pool to create. This struct should contain a static int
+         *  member that sets the number of threads that should be allocated to this pool.
+         *  @code
+         *  struct ThreadPool {
+         *      static const int thread_count = 2;
```
Comment (TrentHouliston) [inline src/dsl/word/Pool.hpp:61]:
```suggestion
         *      static constexpr int thread_count = 2;
```

File: src/dsl/word/Pool.hpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+         *  };
+         *  @endcode
+         *  While it is valid to have a non-const static thread count in the struct, NUClear will not look at any
+         *  changes to this value. Only the first value that NUClear sees will be used to initiate the thread pool.
+         */
+        template <typename PoolType>
+        struct Pool {
+
+            static_assert(PoolType::thread_count > 0, "Can not have a thread pool with less than 1 thread");
```
Comment (TrentHouliston) [inline src/dsl/word/Pool.hpp:70]: This line actually ensures that the thread count is knowable at compile time or this static assert will fail.
So your comment above about non-const static isn't needed since it's impossible to do now

File: src/dsl/word/Pool.hpp [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+            static const util::ThreadPoolDescriptor pool_descriptor;
```
Comment (TrentHouliston) [inline src/dsl/word/Pool.hpp:72]: doc string

File: src/dsl/word/Pool.hpp [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+            template <typename DSL>
+            static inline util::ThreadPoolDescriptor pool(threading::Reaction& /*reaction*/) {
```
Comment (TrentHouliston) [inline src/dsl/word/Pool.hpp:75]: Doc string

File: src/threading/TaskScheduler.hpp [hunk 17]
```diff
@@ -19,52 +19,70 @@
...
-     * rules
-     *  are applied to the tasks in the following order.
+     * PRIORITY
+     *      what priority this task should run with
+     *      tasks are ordered by priority -> creation order
+     * POOL
+     *      which thread pool this task should execute in
+     *      0 being execute on the main thread
+     *      1 being the default pool
+     *      Work out how to create other pools later (fold in always into this?)
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.hpp:49]: You worked this out

File: src/threading/TaskScheduler.hpp [hunk 18]
```diff
@@ -19,52 +19,70 @@
...
+     * GROUP
+     *      which grouping this task belongs to for concurrency (default to the 0 group)
+     * CONCURRENCY
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.hpp:52]: This is part of the group section

File: src/threading/TaskScheduler.hpp [hunk 19]
```diff
@@ -19,52 +19,70 @@
...
+     *      only run if there are less than this many tasks running in this group
+     * INLINE
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.hpp:54]: You called this immediate

File: src/threading/TaskScheduler.hpp [hunk 20]
```diff
@@ -19,52 +19,70 @@
...
      *  @em Priority
      *  @code Priority<P> @endcode
-     *  When a priority is encountered, the task will be scheduled to execute based on this. If one of the three normal
-     *  options are specified (HIGH, DEFAULT and LOW), then within the specified Sync group, it will run before,
-     * normally
-     *  or after other reactions.
+     *  When a priority is encountered, the task will be scheduled to execute based on this. If one of the three
+     * normal options are specified (HIGH, DEFAULT and LOW), then within the specified Sync group, it will run
+     * before, normally or after other reactions.
      *  @attention Note that if Priority<REALTIME> is specified, the Sync type is ignored (Single is not).
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.hpp:63]: I don't think this is a true statement!

File: src/threading/TaskScheduler.cpp [hunk 21]
```diff
@@ -18,62 +19,231 @@
 
 #include "TaskScheduler.hpp"
 
+#include <algorithm>
+#include <atomic>
+#include <iomanip>
+#include <iostream>
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:25]: Doubt you use these in the final implementation

File: src/threading/TaskScheduler.cpp [hunk 22]
```diff
@@ -18,62 +19,231 @@
...
+            run_task(get_task(pool.pool_id));
+        }
+    }
+
+    TaskScheduler::TaskScheduler() {
+        // Setup everything (thread pool, task queue, mutex, condition variable) for the main thread here
+        pools[util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID] =
+            util::ThreadPoolDescriptor{util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID, 1};
+        pool_map[std::this_thread::get_id()] = util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID;
+        queue.emplace(util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID, std::vector<std::unique_ptr<ReactionTask>>{});
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:78]: Did you still want to go back to a single queue or split queue work fine

File: src/threading/TaskScheduler.cpp [hunk 23]
```diff
@@ -18,62 +19,231 @@
...
+        started.store(true);
+
+        // Start all our threads
+        for (const auto& pool : pools) {
+            start_threads(pool.second);
+        }
+
+        // Run main thread tasks
+        pool_func(pools.at(util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID));
+
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:133]: I would put a comment here to say that when the main thread reaches this point it's because the powerplant has shutdown and the main thread is leaving the task scheduler

File: src/threading/TaskScheduler.cpp [hunk 24]
```diff
@@ -18,62 +19,231 @@
...
+            const std::lock_guard<std::mutex> group_lock(queue_mutex);
+            const uint64_t group_id = task->group_descriptor.group_id;
+            if (groups.count(group_id) == 0) {
+                groups.emplace(group_id, 0);
+            }
+        }
+
+        // Check to see if this task was the result of `emit<Direct>`
+        // Direct tasks can run after shutdown and before starting, provided they can be run immediately
+        if (task->immediate) {
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:180]: I think it's a little trickier here, do we want to only care about group and not pool?
E.g. if a MainThread task emits direct we want it to run.
That happens a lot on startup e.g. on<Configuration> extensions do an immediate execution during their construction which is all on the main thread. For those to still work pool needs to be ignored when immediate is true.

We should also make a list of all these things as we define them so we can make unit tests for them

File: src/threading/TaskScheduler.cpp [hunk 25]
```diff
@@ -18,62 +19,231 @@
...
+                run_task(std::move(task));
+                return;
             }
         }
 
-        // Notify a thread that it can proceed
-        condition.notify_one();
+        // We do not accept new tasks once we are shutdown
+        if (running.load()) {
+            const std::lock_guard<std::mutex> queue_lock(queue_mutex);
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:200]: May as well just acquire this lock one line up as we just acquire it again as soon as we leave the if

File: src/threading/TaskScheduler.cpp [hunk 26]
```diff
@@ -18,62 +19,231 @@
...
+
+            // Find where to insert the new task to maintain task order
+            auto it = std::lower_bound(queue.at(current_pool.pool_id).begin(),
+                                       queue.at(current_pool.pool_id).end(),
+                                       task,
+                                       std::less<>());
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:206]: isn't std::less the default here?

File: src/threading/TaskScheduler.cpp [hunk 27]
```diff
@@ -18,62 +19,231 @@
...
+
+            // Insert before the found position
+            queue.at(current_pool.pool_id).insert(it, std::forward<std::unique_ptr<ReactionTask>>(task));
```
Comment (TrentHouliston) [inline src/threading/TaskScheduler.cpp:209]: that forward should be a move

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

File: src/dsl/word/Always.hpp [hunk 28]
```diff
@@ -51,35 +59,113 @@ namespace dsl {
...
+                // Don't emit stats for the idle reaction
+                idle_reaction->emit_stats = false;
+
+                // Keep this reaction handy so it doesn't go out of scope
+                reaction_store[always_reaction->id] = {always_reaction, idle_reaction};
+
+                // Create an unbinder for the always reaction
+                always_reaction->unbinders.push_back([](threading::Reaction& r) {
+                    r.enabled = false;
+                    reaction_store.erase(r.id);
```
[the remark this answers is not in the document]
Comment (Bidski) [reply src/dsl/word/Always.hpp:146]: I don't disagree about cleaning up the thread pool, but we join on this thread on shutdown and provided the actual reaction behaves properly it will get destroyed like all of the other threads and pool

File: src/threading/TaskScheduler.cpp (hunk 22 above)
Comment (Bidski) [reply src/threading/TaskScheduler.cpp:78]: I think split queue is working fine? Unless you think this may have some impact on threads not joining properly?

File: src/threading/TaskScheduler.cpp (hunk 24 above)
Comment (Bidski) [reply src/threading/TaskScheduler.cpp:180]: Yea Im not sure what we want to do with immediate tasks. All of the current tests pass (log tests are probably more important for this), but we also dont have much of a test for an extension and how that would work, except for IOController and networking (the networking test also seems to work)

**Review by TrentHouliston — APPROVED** [review]
